### DISCUSSION TITLE
Is Neo4j Capable of Reasoning from Ontology?

### DISCUSSION THREAD

**Discussion by arkanalexei** [neo4j/neo4j-python-driver#1026] [Q&A]
Hi, sorry if this is not the right place to ask such questions. As the title suggests, is it possible for me to enforce a set of rules from an ontology (.owl for example) and apply that to my Neo4j graph database. So, for example I define an ontology in Protege that has several classes and properties. And then let's say I want to use Neo4j to store the instances. Is there a way to make sure that whatever node or relationship I created follows the ontology?  If it is possible, how can I handle inconsistencies, like creating a node that doesn't follow the defined ontology?

Thank you

**Comment by jbarrasa** [comment]
Hi @arkanalexei , you certainly can. Check out the following resources:
https://github.com/jbarrasa/goingmeta#2022-sessions
**Episode 5: Ontology-driven Knowledge Graph construction**: use an ontology to create a data ingestion pipeline creating a graph that aligns with the domain described by the onto.
**Episode 3: Controlling the shape of your graph with SHACL**: use SHACL to define constraints to control the shape of your graph.

and since you're there, you may want to have a look at other uses of ontologies in Neo4j in other episodes of the series.

**Reply by arkanalexei** [reply]
Interesting resource! However, in the code, I saw that the way you approach this is to import the ontology and translate that into a Cypher query that creates the constraint right? Hm.. I was wondering if there was a cleaner method like how in Ontotext Graph DB you can import the ontology directly and set that as constraints for the database.
